crates/ruff_db/src/vendored.rs

```diff
@@ -0,0 +1,379 @@
+use std::cell::{RefCell, RefMut};
+use std::io::{self, Read};
+use std::ops::{Deref, DerefMut};
+use std::sync::{Mutex, MutexGuard};
+
+use itertools::Itertools;
+use once_cell::sync::Lazy;
+use zip::{read::ZipFile, ZipArchive, ZipWriter};
+
+pub use path::{VendoredPath, VendoredPathBuf};
+
+pub mod path;
+
+type Result<T> = io::Result<T>;
+
+#[derive(Debug)]
+pub struct VendoredFileSystem {
+    inner: VendoredFileSystemInner,
+}
+
+impl Default for VendoredFileSystem {
+    fn default() -> Self {
+        static DATA: Lazy<Box<[u8]>> = Lazy::new(|| {
+            let mut buffer = Vec::new();
+            let cursor = io::Cursor::new(&mut buffer);
+            {
+                let mut archive = ZipWriter::new(cursor);
+                archive.finish().unwrap();
+            }
+            buffer.into_boxed_slice()
+        });
+        Self::new(&DATA).unwrap()
+    }
+}
+
+impl VendoredFileSystem {
+    pub fn new(raw_bytes: &'static [u8]) -> Result<Self> {
+        Ok(Self {
+            inner: VendoredFileSystemInner::new(raw_bytes)?,
+        })
+    }
+
+    pub fn file_type(&self, path: &VendoredPath) -> Option<FileType> {
+        let normalized = normalize_vendored_path(path);
+        let inner_locked = self.inner.lock();
+        let mut archive = inner_locked.borrow_mut();
+        if let Ok(zip_file) = archive.lookup_path(&normalized) {
+            if zip_file.is_dir() {
+                return Some(FileType::Directory);
+            }
+            return Some(FileType::File);
+        }
+        if let Ok(zip_file) = archive.lookup_path(&normalized.with_trailing_slash()) {
+            if zip_file.is_dir() {
+                return Some(FileType::Directory);
+            }
+            return Some(FileType::File);
+        }
+        None
+    }
+
+    pub fn exists(&self, path: &VendoredPath) -> bool {
+        let normalized = normalize_vendored_path(path);
+        let inner_locked = self.inner.lock();
+        let mut archive = inner_locked.borrow_mut();
+        archive.lookup_path(&normalized).is_ok()
+            || archive
+                .lookup_path(&normalized.with_trailing_slash())
+                .is_ok()
+    }
+
+    /// Return the crc32 hash of the file
+    pub fn file_hash(&self, path: &VendoredPath) -> Option<u32> {
+        let normalized = normalize_vendored_path(path);
+        let inner_locked = self.inner.lock();
+        let mut archive = inner_locked.borrow_mut();
+        if let Ok(zip_file) = archive.lookup_path(&normalized) {
+            return Some(zip_file.crc32());
+        }
+        archive
+            .lookup_path(&normalized.with_trailing_slash())
+            .map(|zip_file| zip_file.crc32())
+            .ok()
+    }
```

Comment:
(As discussed offline, we probably are going to need more than these two APIs, since you can have subpackages and namespace packages inside the stdlib just like in any other Python directories)